### PR TITLE
[CodeCompletion] Migrate labeled trailing closure completions to solver-based

### PR DESCRIPTION
**Pull request by ahoppen**
In a first commit, I split `deliverResults` into a `collectResults` call that gathers all the code completion results and then `trySolverCompletion` is responsible for delivering the results to the code completion consumer. 

When we are performing code completion after a call that already has a trailing closure, we first perform the normal postfix completion, which will give us results for operators and member access. We then rewrite the call slightly to include the code completion token as an additional argument, and run argument completion on it. This gives us the completion for the trailing closure label. Once we have collected the results from both completion kinds, we deliver them to the consumer

As a little side-quest, I realized that when computing the code completion results from the cached modules, we were always using the type context of the last lookup that we performed. I changed that slightly to compute a unified type context that contains all possible types from all constraint system solutions. We then use that unified type context to compute the type relations of the global results – this didn’t change any tests.

Oh, and one more thing: When completing after `Void`, we were providing `==`, `<=` etc operators that are defined on tuples. These also showed up when completing trailing closure labels of a call that returns `Void`. Since they are never useful, I now hide operators from code completion if the left-hand-side is `Void`.

rdar://113472967

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please SourceKit stress test

File: lib/IDE/ArgumentCompletion.cpp [hunk 1]
```diff
@@ -50,6 +50,24 @@ bool ArgumentTypeCheckCompletionCallback::addPossibleParams(
     const AnyFunctionType::Param *TypeParam = &ParamsToPass[Idx];
     bool Required = !Res.DeclParamIsOptional[Idx];
 
+    if (Res.FirstTrailingClosureIndex && Idx > *Res.FirstTrailingClosureIndex &&
+        !TypeParam->getPlainType()
+             ->lookThroughAllOptionalTypes()
+             ->is<AnyFunctionType>()) {
+      // We are completing an argument after the first trailing closure, i.e.
+      // a mulitple trailing closure label but the parameter is not a function
```
Comment (bnbarham) [inline lib/IDE/ArgumentCompletion.cpp:58]: `multiple`

File: test/IDE/complete_multiple_trailingclosure.swift [hunk 2]
```diff
@@ -1,31 +1,13 @@
...
+// RUN: %target-swift-ide-test -batch-code-completion -source-filename %s -filecheck %raw-FileCheck -completion-output-dir %t
 
 func globalFunc1(fn1: () -> Int, fn2: () -> String) {}
 func testGlobalFunc() {
   globalFunc1()
     { 1 } #^GLOBALFUNC_SAMELINE^#
     #^GLOBALFUNC_NEWLINE^#
-// GLOBALFUNC_SAMELINE: Begin completions, 1 items
+// GLOBALFUNC_SAMELINE: Begin completions, 2 items
+// GLOBALFUNC_SAMELINE-DAG: Keyword[self]/CurrNominal:          .self[#Void#]; name=self
```
Comment (bnbarham) [inline test/IDE/complete_multiple_trailingclosure.swift:10]: Want to file an issue for removing `self` in ... basically all cases other than static types?

Comment (ahoppen) [reply test/IDE/complete_multiple_trailingclosure.swift:10]: Filed rdar://114322824

**Comment by ahoppen** [comment]
Note to self: The stress tester hit 4 new timeouts, which we are accepting because of the improvements this brings and because it unblocks https://github.com/apple/swift/pull/68080, which deletes a huge amount of code.

**Review by hamishknight — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Pull request merged**